src/core/trulens/core/database/snowparkdb.py

```diff
@@ -0,0 +1,467 @@
+
+        ex = stmt.execute().unique()
+        # unique needed for joinedload above.
+
+        records = [rec[0] for rec in ex]
+        # TODO: Make the iteration of records lazy in some way. See
+        # TODO(piotrm) above.
+
+        return AppsExtractor().get_df_and_cols(records=records)
+
+    def _escape_single_quotes(self, value):
+        if isinstance(value, str):
+            return value.replace("'", "''")
+        return value
+
+    def _create_updatesql(self, orm_class: mod_orm.ORM, new_obj, key: str, key_value: str):
+        # Collect the row as a dictionary
+        row_dict = {k: self._escape_single_quotes(v) for k, v in new_obj.__dict__.items() if not k.startswith('_') and v is not None}
+
+        # Generate the SET clause for the SQL update statement
+        set_clause = ", ".join([f"{k}='{v}'" for k, v in row_dict.items() if k != key])
+        sql_update = f"UPDATE {orm_class.__tablename__} SET {set_clause} WHERE {key}='{key_value}'"
+        return sql_update
+
+    def _insert_sql(self, orm_class: mod_orm.ORM, new_obj):
+        # Collect the row as a dictionary
+        row_dict = {k: self._escape_single_quotes(v) for k, v in new_obj.__dict__.items() if not k.startswith('_') and v is not None}
+        # Generate the INSERT INTO clause for the SQL insert statement
+        columns = ", ".join(row_dict.keys())
+        values = ", ".join([f"'{v}'" for v in row_dict.values()])
+        sql_insert = f"INSERT INTO {orm_class.__tablename__} ({columns}) VALUES ({values})"
+        return sql_insert
+
+    def insert_feedback(
+        self, feedback_result: mod_feedback_schema.FeedbackResult
+    ) -> mod_types_schema.FeedbackResultID:
+        """See [DB.insert_feedback][trulens.core.database.base.DB.insert_feedback]."""
+        _feedback_result = self.orm.FeedbackResult.parse(
+            feedback_result
+        )
+        feedback_results_table = self.session.table(self.orm.FeedbackResult.__tablename__)
+        row_exists = feedback_results_table.filter(
+            col('feedback_result_id') == _feedback_result.feedback_result_id
+        ).first()
+    
+        if row_exists:
+            sql = self._create_updatesql(
+                self.orm.FeedbackResult,
+                _feedback_result,
+                'feedback_result_id',
+                _feedback_result.feedback_result_id)
+            res = self.session.sql(sql).collect()
+        else:
+            res = self.session.sql(self._insert_sql(self.orm.FeedbackResult, _feedback_result)).collect()
+        
+        status = mod_feedback_schema.FeedbackResultStatus(
+            _feedback_result.status
+        )
+        logger.info(
+            "%s feedback result %s %s %s",
+            res,
+            _feedback_result.name,
+            status.name,
+            _feedback_result.feedback_result_id,
+        )
+        return _feedback_result.feedback_result_id
+
+    def batch_insert_feedback(self):
+        pass
```

Comment:
Can you make these `raise NotImplementedError(...)` to make sure we can catch them if they ever get called.